nestkernel/input_backend_mpi.cpp

```diff
@@ -0,0 +1,299 @@
+    device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
+    if ( device_it != devices_[ tid ].end() )
+    {
+      devices_[ tid ].erase( device_it );
+    }
+    std::pair< MPI_Comm*, InputDevice* > pair = std::make_pair( nullptr, &device );
+    devices_[ tid ].insert( std::make_pair( node_id, pair ) );
+  }
+  else
+  {
+    throw BadProperty( "Only spike generators can have input backend 'mpi'." );
+  }
+}
+
+void
+nest::InputBackendMPI::disenroll( InputDevice& device )
+{
+  thread tid = device.get_thread();
+  index node_id = device.get_node_id();
+
+  device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
+  if ( device_it != devices_[ tid ].end() )
+  {
+    devices_[ tid ].erase( device_it );
+  }
+}
+
+void
+nest::InputBackendMPI::set_value_names( const InputDevice& device,
+  const std::vector< Name >& double_value_names,
+  const std::vector< Name >& long_value_names )
+{
+  // nothing to do
+}
+
+void
+nest::InputBackendMPI::prepare()
+{
+  // Create the connection with MPI by thread ( I am not sure is the best)
+  // 1) take all the port of the connections
+  thread thread_id = kernel().vp_manager.get_thread_id();
+
+  // get port and update the list of device
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
+  {
+    // add the link between MPI communicator and the device (device can share the same MPI communicator
+    std::string port_name;
+    get_port( it_device->second.second, &port_name );
+    comm_map::value_type::iterator comm_it = commMap_[ thread_id ].find( port_name );
+    MPI_Comm* comm;
+    if ( comm_it != commMap_[ thread_id ].end() )
+    {
+      comm = comm_it->second.first;
+      comm_it->second.second += 1;
+    }
+    else
+    {
+      comm = new MPI_Comm;
+      std::pair< MPI_Comm*, int > comm_count = std::make_pair( comm, 1 );
+      commMap_[ thread_id ].insert( std::make_pair( port_name, comm_count ) );
+    }
+    it_device->second.first = comm;
+  }
+
+  // 2) connect the thread with MPI process it need to be connected
+  // WARNING can be a bug if it's need all the thread to be connected in MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
+  {
+    MPI_Comm_connect( it_comm->first.data(),
+      MPI_INFO_NULL,
+      0,
+      MPI_COMM_WORLD,
+      it_comm->second.first ); // should use the status for handle error
+    printf( "Connect to %s\n", it_comm->first.data() );
+    fflush( stdout );
+  }
+}
+
+void
+nest::InputBackendMPI::pre_run_hook()
+{
+  // connect take information of MPI process (for the moment only spike train)
+  const thread thread_id = kernel().vp_manager.get_thread_id();
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```

Comment:
Use `auto`, or a range-based for loop if possible.

Reply:
done